### ISSUE TITLE
Connection to etcd cluster isn't resilient

### ISSUE THREAD

**Issue by balena** [zhongwencool/eetcd#49]
Steps to reproduce:

1. Connect to `etcd`;
2. Interrupt the connectivity, and wait for all connections to timeout/disconnect;
3. Restablish connectivity.

**Expected:** `eetcd` reconnects to etcd cluster.

**Behavior:** it doesn't reconnect.

Instead, it loses the configuration initially passed at the `eetcd:open` function.

If you use `auto_sync_interval_ms` option, the following logs are generated, repeatedly:
```
[warning] <Name> has no active connections to etcd cluster, cannot get member list
[warning] <Name> has no active connections to etcd cluster, cannot get member list
...
```

At the client app, the error `:eetcd_conn_unavailable` is constantly returned.

**Comment by michaelklishin** [comment]
It can be a reasonable design choice for a client library to not perform connection recovery,
in particular in Erlang, and let the developer do that.

I am not sure if that's the case here. @zhongwencool would you be interested in having this feature?

**Comment by balena** [comment]
Attempted to address the problem in https://github.com/zhongwencool/eetcd/pull/50.

> It can be a reasonable design choice for a client library to not perform connection recovery
Yes, it can, but in this case it must exist a way for the client to discover that the connection is down, and provide a function to reconnect, which is not the case.

The PR solves the problem by reusing the health checks to refill the `freeze_conns` attribute from `members_list` and thus restablish the connections.

I couldn't understand the design decisions behind of erasing active connections and freeze connections at the same time. This means when both are empty we get a dangling connection...

**Comment by cjbottaro** [comment]
Confirming this issue. Our Etcd cluster ran out of memory, so we had to restart it. The app (which uses this library) did not recover on its own. We had to restart the app pods to get working again.

I'm not sure how gRPC stuff works, but most network clients that I write use the awesome `Connection` module/behaviour which makes it super easy to manage reconnections for `:gen_tcp` sockets and what not.

**Comment by michaelklishin** [comment]
@cjbottaro what version of this library do you use? It should have been addressed by https://github.com/balena/eetcd/commit/b7d92a344eec4521fd2650d1815e347ec787933a.

**Comment by cjbottaro** [comment]
`mix deps` reports `0.3.5`...
```
* eetcd 0.3.5 (Hex package) (rebar3)
  locked at 0.3.5 (eetcd) af9d5158
```

Unfortunately, I didn't save the app logs when we restarted the Etcd cluster, but there were definitely messages from `eetcd`.

**Comment by michaelklishin** [comment]
There will be messages from `eetcd` since connection loss will result either in an error somewhere or just an error log message. The question is, how do we reproduce the failure to recover.

@cjbottaro can you put together a small app that demonstrates the non-recovery behavior and how your application supervises whatever processes use `eetcd`?

**Comment by balena** [comment]
It's easy to reproduce. The steps were described at the top of this issue.

**Comment by cjbottaro** [comment]
Yup, super easy to reproduce with Docker. In our case Kubernetes... we just deleted our cluster:
```
00:31:53.147 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.159 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.276 [warning] Etcd failed to connect [etcd-2:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.313 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.318 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.326 [warning] Etcd failed to connect [etcd-2:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:53.487 [warning] Etcd failed to connect [etcd-1:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:54.488 [warning] Etcd failed to connect [etcd-2:2379] by <Gun Down> :timeout

00:31:54.493 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:54.496 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:31:55.497 [warning] Etcd failed to connect [etcd-2:2379] by <Gun Down> :timeout

00:31:56.498 [warning] Etcd failed to connect [etcd-1:2379] by <Gun Down> :timeout

00:31:58.300 [warning] Etcd failed to connect [etcd-1:2379] by <Gun Down> :timeout

00:31:59.301 [warning] Etcd failed to connect [etcd-2:2379] by <Gun Down> :timeout

00:31:59.305 [warning] Etcd failed to connect [etcd-0:2379] by <Gun Down> {:shutdown, :econnrefused}

00:32:00.905 [info] ETCD(Etcd, #PID<0.5673.0>)'s connections are ready.
```

That last log message is super suspect, considering the cluster (including DNS) doesn't exist anymore at all.

**Comment by michaelklishin** [comment]
This has not been forgotten but I personally won't be able to look into this in the near term.

**Comment by gilbertwong96** [comment]
> It can be a reasonable design choice for a client library to not perform connection recovery, in particular in Erlang, and let the developer do that.
> 
> I am not sure if that's the case here. @zhongwencool would you be interested in having this feature?

@michaelklishin I suppose it isn't a reasonable design choice. This problem only occurs when the connect mode is set to `connect_all` by default. The connection recovery mechanism works if you set the `connect mode` to `random`.

**Comment by michaelklishin** [comment]
Good to know that it only affects one mode, thanks for digging in.

**Comment by gilbertwong96** [comment]
> Good to know that it only affects one mode, thanks for digging in.

@michaelklishin I fixed in #54.

**Comment by zhongwencool** [comment]
`0.3.6` is ready now.
